handlers/auth_handler.go

```diff
@@ -472,3 +472,237 @@ func LogoutHandler(c echo.Context) error {
+// @Router       /v1/auth/reset-password [post]
+func ResetPasswordHandler(c echo.Context) error {
+	logger := c.Logger()
+
+	var req ResetPasswordRequest
+	if err := c.Bind(&req); err != nil {
+		logger.Error("Invalid password reset request payload:", err)
+		return echo.ErrBadRequest
+	}
+
+	if req.Token == "" {
+		logger.Error("Password reset token is required")
+		return &echo.HTTPError{
+			Code:    http.StatusBadRequest,
+			Message: "token field is required",
+		}
+	}
+
+	if req.NewPassword == "" {
+		logger.Error("New password is required")
+		return &echo.HTTPError{
+			Code:    http.StatusBadRequest,
+			Message: "new_password field is required",
+		}
+	}
+
+	if err := passwordcheck.ValidatePassword(c.Request().Context(), req.NewPassword); err != nil {
+		logger.Error("New password validation failed: ", err)
+		return &echo.HTTPError{
+			Code:    http.StatusBadRequest,
+			Message: "Invalid new password: " + err.Error(),
+		}
+	}
+
+	passwordReset := models.PasswordReset{}
+
+	if err := db.Conn.Preload("User").
+		Where("token = ? AND is_used = ?", req.Token, false).
+		First(&passwordReset).Error; err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			logger.Error("Invalid or already used password reset token")
+			return &echo.HTTPError{
+				Code:    http.StatusBadRequest,
+				Message: "Invalid or already used password reset token",
+			}
+		}
+		logger.Errorf("Failed to find password reset record: %v", err)
+		return echo.ErrInternalServerError
+	}
+
+	if time.Now().After(passwordReset.ExpiresAt) {
+		logger.Error("Password reset token has expired")
+		return &echo.HTTPError{
+			Code:    http.StatusGone,
+			Message: "Password reset token has expired. Please request a new one.",
+		}
+	}
+
+	newCrypto := crypto.NewCrypto()
+
+	if err := newCrypto.VerifyPassword(req.NewPassword, passwordReset.User.Password); err == nil {
+		logger.Error("New password is the same as current password.")
+		return &echo.HTTPError{
+			Code:    http.StatusBadRequest,
+			Message: "New password must be different from the current password",
+		}
+	}
+
+	hashedNewPassword, err := newCrypto.HashPassword(req.NewPassword)
+	if err != nil {
+		logger.Errorf("Failed to hash new password: %v", err)
+		return echo.ErrInternalServerError
+	}
+
+	tx := db.Conn.Begin()
+
+	if err := tx.Model(&passwordReset.User).Update("password", hashedNewPassword).Error; err != nil {
+		tx.Rollback()
+		logger.Errorf("Failed to update user password: %v", err)
+		return echo.ErrInternalServerError
+	}
+
+	if err := tx.Model(&passwordReset).Update("is_used", true).Error; err != nil {
+		tx.Rollback()
+		logger.Errorf("Failed to mark password reset token as used: %v", err)
+		return echo.ErrInternalServerError
+	}
+
+	if err := tx.Where("user_id = ?", passwordReset.User.ID).Delete(&models.Session{}).Error; err != nil {
+		tx.Rollback()
+		logger.Errorf("Failed to invalidate user sessions: %v", err)
+		return echo.ErrInternalServerError
+	}
+
+	tx.Commit()
```

Comment:
We should use `models.EmailVerification{}` instead of `models.PasswordReset{}`

Reply:
We should use more inline error handling from the beginning of the transaction to committing the transaction.